File: src/stack/StackOnArray.go
package stack

import "sync"

// ArrayStack is a stack based on array
// which means the stack is a continious memory space
// we need tp dynamically extend the memory to avoid stack overload
type ArrayStack struct {
	data []interface{}
	top  int
	//considering the concurrent, introduce the Mutex
	//also can use RWMutx according to the contax
	lock sync.Mutex
}

//NewArrayStack constructor
func NewArrayStack() *ArrayStack {
	return &ArrayStack{
		data: make([]interface{}, 0, 32),
		top:  -1,
		lock: sync.Mutex{},
	}
}

//IsEmpty return true is the stack is empty
func (as *ArrayStack) IsEmpty() bool {
	as.lock.Lock()
	defer as.lock.Unlock()

	return as.top < 0
}

//Push stack push
func (as *ArrayStack) Push(item interface{}) {
	as.lock.Lock()
	defer as.lock.Unlock()

	if as.top < 0 {
		as.top = 0
	} else {
		as.top++
	}

	if as.top > len(as.data)-1 {
		as.data = append(as.data, item)
	} else {
		as.data[as.top] = item
	}
}

//Pop stack pop
func (as *ArrayStack) Pop() interface{} {
	as.lock.Lock()
	defer as.lock.Unlock()

	if as.IsEmpty() {
		return nil
	}

	v := as.data[as.top]
	as.top--
	return v

}

//Top stack top
func (as *ArrayStack) Top() interface{} {
	as.lock.Lock()
	defer as.lock.Unlock()

	if as.IsEmpty() {
		return nil
	}
	return as.data[as.top]
}

//Flush stack flush
func (as *ArrayStack) Flush() {
	as.lock.Lock()
	defer as.lock.Unlock()

	as.top = -1
}

File: src/stack/StackOnArray.java
package stack;


class ArrayStack<T> {
    private T[] data;
    private int top;
    private int capacity;

    private static int d4tCap = 32;

    public ArrayStack(final int capa) {
        data = (T[]) new Object[capa];
        capacity = capa;
        top = -1;
    }

    public ArrayStack() {
        data = (T[]) new Object[d4tCap];
        capacity = d4tCap;
        top = -1;
    }

    public void Push(final T item) {
        if (top < 0) {
            top = 0;
        }

        if (top == capacity - 1) {
            resize();
        }

        top++;
        data[top] = item;

    }

    public T Pop() {
        if (top <= 0) {
            return null;
        }

        final T v = data[top];
        top--;
        return v;
    }

    public void Flush() {
        top = -1;

    }

    public T Top() {
        return (top > -1) ? data[top] : null;
    }

    public void resize() {
        T[] newdata = (T[]) new Object[capacity*2];

        for (int i = 0 ; i < capacity; i ++) {
            newdata[i] = data[i];
        }

        capacity = capacity*2;
        data = newdata;
    }
}